### PR TITLE
react-syntax-highlighter: version 12.2

### PR DESCRIPTION
**Pull request by guoyunhe**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by sandersn** [comment]
@guoyunhe Thanks for the update to react-syntax-highlighter. I haven't been able to review this large change, and in addition I don't like merging changes to such a widely used package without some authors signing off.

It would help to revert prettier's formatting changes and submit them in a separate PR.

 @NoHomey @ajgamble-milner @anirban09 
Can you take a look?

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
Needs experts sign off

**Comment by haysclark** [comment]
@guoyunhe I'm interested in Types for `react-syntax-highlighter` 12.x. Any chance you are following up on this PR?

File: types/react-syntax-highlighter/index.d.ts [hunk 1]
```diff
@@ -9,17 +9,39 @@
...
-        style?: any;
-        customStyle?: any;
-        lineProps?: lineTagPropsFunction | React.HTMLProps<HTMLElement>;
+        style?: Style;
+        children?: string;
+        customStyle?: React.CSSProperties;
         codeTagProps?: React.HTMLProps<HTMLElement>;
         useInlineStyles?: boolean;
         showLineNumbers?: boolean;
+        showInlineLineNumbers?: boolean;
```
Comment (Jessidhia) [inline types/react-syntax-highlighter/index.d.ts:35]: This is a 13.0 feature, not 12.2; need to update the version header.

File: types/react-syntax-highlighter/index.d.ts [hunk 2]
```diff
@@ -9,17 +9,39 @@
...
+
+interface RendererParams {
+    rows: any;
+    stylesheet?: any;
+    useInlineStyles?: boolean;
+}
+
+type Renderer = (params: RendererParams) => React.ReactNode;
+
 declare module "react-syntax-highlighter" {
```
Comment (Jessidhia) [inline types/react-syntax-highlighter/index.d.ts:26]: If this is a 13.0 definition (like the presence of `showInlineLineNumbers` implies), this is missing the `createElement` export

File: types/react-syntax-highlighter/index.d.ts [hunk 3]
```diff
@@ -44,15 +66,15 @@ declare module "react-syntax-highlighter" {
 // esm start
 declare module "react-syntax-highlighter/dist/esm/default-highlight" {
     import * as React from "react";
-    import { SyntaxHighlighterProps } from 'react-syntax-highlighter';
+    import { SyntaxHighlighterProps } from "react-syntax-highlighter";
     export default class SyntaxHighlighter extends React.Component<
```
Comment (Jessidhia) [inline types/react-syntax-highlighter/index.d.ts:70]: In 13.0 these components are now function components (not very important other than forbidding refs)

File: types/react-syntax-highlighter/index.d.ts [hunk 4]
```diff
@@ -9,17 +9,39 @@
...
-        lineNumberStyle?: any;
+        lineNumberContainerStyle?: React.CSSProperties;
+        lineNumberStyle?: React.CSSProperties;
+        wrapLines?: boolean;
+        lineProps?: lineTagPropsFunction | React.HTMLProps<HTMLElement>;
+        renderer?: Renderer;
+        PreTag?: any;
+        CodeTag?: any;
+
         [spread: string]: any;
```
Comment (Jessidhia) [inline types/react-syntax-highlighter/index.d.ts:45]: The `...rest` from the component is mixed in with the props for the `PreTag`.

A more complete definition, without using this `[string]:any` index, could be something like:

```ts
interface SyntaxHighlighterProps<T extends React.ElementType = 'pre'> extends
  React.ComponentPropsWithoutRef<T> {
  // ...
  PreTag?: T
  // ...
}
```

**Review by Jessidhia — COMMENTED** [review]
- Most of the thousands of lines deleted were really individual declarations of each theme, which got simplified into `/*` declarations. This does miss out on some type checking benefits but it's probably not a problem.
- The only formatting changes are changes from single quotes to double quotes, and removal of some newlines in the re-exporting module.

**Comment by bsgreenb** [comment]
Any update on this?  Or do we have to drop in a d.t.s workaround still?

**Pull request closed without merging**